src/main/java/com/mongodb/hibernate/translate/attachment/Attachment.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.translate.attachment;
+
+import static java.lang.String.format;
+
+import com.mongodb.hibernate.internal.VisibleForTesting;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Manages global state while translating SQL AST via {@link org.hibernate.sql.ast.SqlAstWalker}.
+ *
+ * <p>{@code SqlAstWalker} is based on {@code visitor} pattern without returning value from the various visitor methods.
+ * The original goal of {@code SqlAstWalker} is for concatenating SQL string internally; but when translating SQL AST
+ * into MQL, it is highly desirable to return value (e.g. {@link com.mongodb.hibernate.translate.mongoast.AstNode}). To
+ * avoid such dilemma, some attachment/detachment mechanism is implemented as follows:
+ *
+ * <ul>
+ *   <li>when some visitor method needs to return some value, it attaches the value by invoking
+ *       {@link #attach(AttachmentKey, Object)}.
+ *   <li>when some coordinating detacher needs to access the value attached above, it invokes
+ *       {@link #expect(AttachmentKey, Runnable)}; internally the {@code Runnable} will attach value by the above
+ *       attacher.
+ *   <li>both attacher and detacher should provide identical {@link AttachmentKey} to ensure data transferring safety.
+ * </ul>
+ *
+ * @see org.hibernate.sql.ast.SqlAstWalker
+ * @see AttachmentKey
+ */
+@SuppressWarnings("unchecked")
+public final class Attachment {
```

Comment:
1. I think, the "attachment" term is a misnomer here. It came from `RetryState.attachment`, but here the idea is very different, despite the implementation being similar. `Attachment` exists to allow "returning" a value from a `void` method in a situation when we can't make it non-`void`. The fact that this is implemented by writing that value into the instance field of `Attachment` (which one may see as "attaching") is not of importance, certainly not on importance to the extent that it dictates the naming.
2. This mechanism, allowing us to return values from `void` methods, is not coupled with the SQL AST walker, so I propose us to remove the story about the walker, which is essentially the whole class-level documentation of `Attachment`.

Reply:
Now I read the simulated stuff in Java driver (mostly written by you), I agree `Attachment` is inappropriate here. It was used for storing retry state related common states. In our usage, we have different requirements, including:

1. we store one single value for the visitor method returning void in question; attachment is barely feasible in the sense that we did store value somewhere
2. A more subtle requirement to restore previous state after finish the current visitor method to accommodate tree depth-first traversal, which is the salient feature and we should make it explicit in doc; it is based on a clever trick that we spread ancestor states within the various method calling stack trace corresponding to the visitor methods of its various ancestors.